src/codecs/framing/character_delimited.rs

```diff
@@ -137,17 +135,47 @@ impl Decoder for CharacterDelimitedDecoder {
     }
 }
 
-impl<T> Encoder<T> for CharacterDelimitedDecoder
-where
-    T: AsRef<[u8]>,
-{
-    type Error = io::Error;
-
-    fn encode(&mut self, item: T, buf: &mut BytesMut) -> Result<(), io::Error> {
-        let item = item.as_ref();
-        buf.reserve(item.len() + 1);
-        buf.put(item);
-        buf.put_u8(self.delimiter as u8);
+/// Config used to build a `CharacterDelimitedEncoder`.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct CharacterDelimitedEncoderConfig {
+    character_delimited: CharacterDelimitedEncoderOptions,
+}
+
+/// Options for building a `CharacterDelimitedEncoder`.
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+pub struct CharacterDelimitedEncoderOptions {
+    /// The character that delimits byte sequences.
+    delimiter: u8,
+}
+
+#[typetag::serde(name = "character_delimited")]
+impl encoding::FramingConfig for CharacterDelimitedEncoderConfig {
+    fn build(&self) -> crate::Result<encoding::BoxedFramer> {
+        Ok(Box::new(CharacterDelimitedEncoder::new(
+            self.character_delimited.delimiter,
+        )))
+    }
+}
+
+/// An encoder for handling bytes that are delimited by (a) chosen character(s).
+#[derive(Debug, Clone)]
+pub struct CharacterDelimitedEncoder {
+    delimiter: u8,
+}
+
+impl CharacterDelimitedEncoder {
+    /// Creates a `CharacterDelimitedEncoder` with the specified delimiter.
+    pub const fn new(delimiter: u8) -> Self {
+        Self { delimiter }
+    }
+}
+
+impl Encoder<()> for CharacterDelimitedEncoder {
+    type Error = encoding::BoxedFramingError;
+
+    fn encode(&mut self, _: (), buf: &mut BytesMut) -> Result<(), encoding::BoxedFramingError> {
+        buf.reserve(buf.len() + 1);
+        buf.put_u8(self.delimiter);
```

Comment:
I believe `reserve` reserves the passed number of _additional_ bytes, not total. So you shouldn't need to add `buf.len()` here.

Reply:
Oh wow, thanks for catching that. TIL that the same is true for `std` data structures. How did I miss this fact until now 😮

Reply:
That `reserve` in particular seems unnecessary either way

Reply:
Probably got that from C++: https://www.cplusplus.com/reference/vector/vector/reserve/, not looking back 😄